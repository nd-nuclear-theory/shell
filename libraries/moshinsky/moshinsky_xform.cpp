/****************************************************************
  moshinsky_xform.cpp

  Mark A. Caprio
  University of Notre Dame

****************************************************************/

#include "moshinsky/moshinsky_xform.h"

#include "moshinsky/moshinsky_bracket.h"

namespace moshinsky {

double RacahReductionFactorFirstSystemGT(
    const HalfInt& J1p, const HalfInt& Jp, 
    const HalfInt& J1, const HalfInt& J, 
    const HalfInt& J2, const HalfInt& J0
  )
{
  double value = ParitySign(J1p+J2+J+J0)
    *Hat(J1p)*Hat(J)
    *am::Wigner6J(J1p,Jp,J2,J,J1,J0);
  return value;
}

////////////////////////////////////////////////////////////////
// Moshinsky transformation
////////////////////////////////////////////////////////////////

Eigen::MatrixXd
MoshinskyMatrixNLSJT(
    const basis::RelativeCMSubspaceNLSJT& relative_cm_subspace,
    const basis::TwoBodySubspaceNLSJT& two_body_subspace
  )
{

  // set up matrix to hold results
  Eigen::MatrixXd matrix(relative_cm_subspace.size(),two_body_subspace.size());

  // guard against call with mismatched subspaces
  assert(relative_cm_subspace.N()==two_body_subspace.N());
  assert(relative_cm_subspace.L()==two_body_subspace.L());
  assert(relative_cm_subspace.S()==two_body_subspace.S());
  assert(relative_cm_subspace.J()==two_body_subspace.J());
  assert(relative_cm_subspace.T()==two_body_subspace.T());

  // retrieve L for sector
  int L = relative_cm_subspace.L();

  // populate matrix
  for (int index_relative_cm=0; index_relative_cm<relative_cm_subspace.size(); ++index_relative_cm)
    for (int index_two_body=0; index_two_body<two_body_subspace.size(); ++index_two_body)
      {

        // retrieve relative-cm state
        basis::RelativeCMStateNLSJT relative_cm_state(relative_cm_subspace,index_relative_cm);
        int Nr = relative_cm_state.Nr();
        int lr = relative_cm_state.lr();
        int nr = (Nr - lr)/2;
        int Nc = relative_cm_state.Nc();
        int lc = relative_cm_state.lc();
        int nc = (Nc - lc)/2;
        int L = relative_cm_state.L();

        // retrieve two-body state
        basis::TwoBodyStateNLSJT two_body_state(two_body_subspace,index_two_body);
        int N1 = two_body_state.N1();
        int l1 = two_body_state.l1();
        int n1 = (N1 - l1)/2;
        int N2 = two_body_state.N2();
        int l2 = two_body_state.l2();
        int n2 = (N2 - l2)/2;

        // evaluate bracket
        matrix(index_relative_cm,index_two_body) = moshinsky::MoshinskyBracket(nr,lr,nc,lc,n1,l1,n2,l2,L);
				  
      }

  // return matrix
  //  std::cout << "Moshinsky " << matrix << std::endl;
  return matrix;
}


Eigen::MatrixXd
RelativeCMMatrixNLSJT(
    const basis::RelativeSpaceLSJT& relative_space,
    const basis::RelativeSectorsLSJT& relative_sectors,
    const basis::MatrixVector& relative_matrices,
    const typename basis::RelativeCMSectorsNLSJT::SectorType& relative_cm_sector,
    int J0
  )
{
  
  // initialize matrix
  Eigen::MatrixXd relative_cm_matrix = Eigen::MatrixXd::Zero(
      relative_cm_sector.ket_subspace().size(),
      relative_cm_sector.bra_subspace().size()
    );

  // extract relative-cm target subspace labels
  // int Np = relative_cm_sector.ket_subspace().N();
  // int Lp = relative_cm_sector.ket_subspace().L();
  // int Sp = relative_cm_sector.ket_subspace().S();
  // int Jp = relative_cm_sector.ket_subspace().J();
  // int Tp = relative_cm_sector.ket_subspace().T();
  // int gp = relative_cm_sector.ket_subspace().g();
  // int N = relative_cm_sector.bra_subspace().N();
  // int L = relative_cm_sector.bra_subspace().L();
  // int S = relative_cm_sector.bra_subspace().S();
  // int J = relative_cm_sector.bra_subspace().J();
  // int T = relative_cm_sector.bra_subspace().T();
  // int g = relative_cm_sector.bra_subspace().g();

  // iterate over spurious relative (target) matrix elements
  for (int relative_cm_bra_index=0; relative_cm_bra_index<relative_cm_sector.bra_subspace().size(); ++relative_cm_bra_index)
    for (int relative_cm_ket_index=0; relative_cm_ket_index<relative_cm_sector.ket_subspace().size(); ++relative_cm_ket_index)
      {

        // extract needed relative-cm state labels
        basis::RelativeCMStateNLSJT relative_cm_bra(relative_cm_sector.bra_subspace(),relative_cm_bra_index);
        int Nrp = relative_cm_bra.Nr();
        int lrp = relative_cm_bra.lr();
        int Ncp = relative_cm_bra.Nc();
        int lcp = relative_cm_bra.lc();
        int Lp = relative_cm_bra.L();
        int Sp = relative_cm_bra.S();
        int Jp = relative_cm_bra.J();
        int Tp = relative_cm_bra.T();

        basis::RelativeCMStateNLSJT relative_cm_ket(relative_cm_sector.ket_subspace(),relative_cm_ket_index);
        int Nr = relative_cm_ket.Nr();
        int lr = relative_cm_ket.lr();
        int Nc = relative_cm_ket.Nc();
        int lc = relative_cm_ket.lc();
        int L = relative_cm_ket.L();
        int S = relative_cm_ket.S();
        int J = relative_cm_ket.J();
        int T = relative_cm_ket.T();

        // short circuit if cm motion not conserved
        if (!((Ncp==Nc)&&(lcp==lc)))
          continue;

        // check for adequate relative N truncation on source interaction
        assert(std::max(Nrp,Nr)<=relative_space.Nmax());

        // determine relative (source) angular momenta
        //   by triangle selection between relative and cm motion
        HalfInt::pair relative_bra_Jr_range = am::AngularMomentumRangeIntersection(
            am::ProductAngularMomentumRange(relative_cm_bra.lc(),relative_cm_bra.J()),
            am::ProductAngularMomentumRange(relative_cm_bra.lr(),relative_cm_bra.S())
          );
        HalfInt::pair relative_ket_Jr_range = am::AngularMomentumRangeIntersection(
            am::ProductAngularMomentumRange(relative_cm_ket.lc(),relative_cm_ket.J()),
            am::ProductAngularMomentumRange(relative_cm_ket.lr(),relative_cm_ket.S())
          );

        // sum relative angular momentum contributions
        double relative_cm_matrix_element = 0;
        for (int Jrp=int(relative_bra_Jr_range.first); Jrp<=int(relative_bra_Jr_range.second); ++Jrp)
          for (int Jr=int(relative_ket_Jr_range.first); Jr<=int(relative_ket_Jr_range.second); ++Jr)
            {

              // short circuit if relative J truncation on source
              // interaction eliminates this source sector
              if (!(std::max(Jrp,Jr)<=relative_space.Jmax()))
                continue;

              // look up relative sector
              int grp = lrp%2;
              int relative_bra_subspace_index = relative_space.LookUpSubspaceIndex(
                  basis::RelativeSubspaceLSJT::SubspaceLabelsType(lrp,Sp,Jrp,Tp,grp)
                );
              int gr = lr%2;
              int relative_ket_subspace_index = relative_space.LookUpSubspaceIndex(
                  basis::RelativeSubspaceLSJT::SubspaceLabelsType(lr,S,Jr,T,gr)
                );
              int relative_sector_index = relative_sectors.LookUpSectorIndex(
                  relative_bra_subspace_index,
                  relative_ket_subspace_index
                );

              // look up relative matrix element
              const basis::RelativeSubspaceLSJT& relative_bra_subspace = relative_space.GetSubspace(
                  relative_bra_subspace_index
                );
              int relative_bra_index = relative_bra_subspace.LookUpStateIndex(
                  basis::RelativeStateLSJT::StateLabelsType(Nrp)
                );
              const basis::RelativeSubspaceLSJT& relative_ket_subspace = relative_space.GetSubspace(
                  relative_ket_subspace_index
                );
              int relative_ket_index = relative_ket_subspace.LookUpStateIndex(
                  basis::RelativeStateLSJT::StateLabelsType(Nr)
                );
              double relative_matrix_element = relative_matrices[relative_sector_index](relative_bra_index,relative_ket_index);
              
              // accumulate contribution to relative-cm matrix element
              relative_cm_matrix_element
                += am::Unitary6JZ(lrp,lc,Lp,Jp,Sp,Jrp) * am::Unitary6JZ(lr,lc,L,J,S,Jr)
                * RacahReductionFactorFirstSystemGT(Jrp,Jp,Jr,J,lr,J0)
                * relative_matrix_element;
            }	

        // save matrix element
        relative_cm_matrix(relative_cm_bra_index,relative_cm_ket_index) = relative_cm_matrix_element;
      }

  // return matrix
  return relative_cm_matrix;

}

Eigen::MatrixXd 
TransformedSector(
    const basis::TwoBodySubspaceNLSJT& two_body_subspace2,
    const basis::TwoBodySubspaceNLSJT& two_body_subspace1,
    const basis::RelativeSpaceLSJT& relative_space,
    const basis::RelativeSectorsLSJT& relative_sectors,
    const basis::MatrixVector& relative_matrices,
    int J0
  )
{
  // define target matrix
  Eigen::MatrixXd A = Eigen::MatrixXd::Zero(two_body_subspace2.size(),two_body_subspace1.size());

  // extract subspace labels
  int L2 = two_body_subspace2.L();
  int S2 = two_body_subspace2.S();
  int J2 = two_body_subspace2.J();
  int T2 = two_body_subspace2.T();
  int g2 = two_body_subspace2.g();

  int L1 = two_body_subspace1.L();
  int S1 = two_body_subspace1.S();
  int J1 = two_body_subspace1.J();
  int T1 = two_body_subspace1.T();
  int g1 = two_body_subspace1.g();


  // obtain N truncations
  //
  // spurious spaces contribute for Nmax only up to lesser of the bra
  // and ket two-body space Nmax values
//  int Nmax2 = two_body_subspace2.Nmax();
//  int Nmax1 = two_body_subspace1.Nmax();
//  int Nc_max = std::min(Nmax2,Nmax1); 

//  // accumulate cm contributions
//  for (int Nc=0; Nc <= Nc_max; ++Nc)
//    for (int lc = Nc%2; lc <= Nc; lc+=2)
//      {
//        // define intermediate spurious relative subspaces
//        const basis::RelativeCMSubspaceLSJT relative_cm_subspace2(Nc,lc,L2,S2,J2,T2,g2,Nmax2);
//        const basis::RelativeCMSubspaceLSJT relative_cm_subspace1(Nc,lc,L1,S1,J1,T1,g1,Nmax1);
//
//        // check that spurious relative subspace is nonempty
//        if (relative_cm_subspace2.size()==0)
//          continue;
//        if (relative_cm_subspace1.size()==0)
//          continue;
//
//        // construct Ar_spurious
//        // PLACEHOLDER: Eigen::MatrixXd As = Eigen::MatrixXd::Zero(relative_cm_subspace2.size(),relative_cm_subspace1.size());
//        Eigen::MatrixXd As = RelativeCMMatrix(
//            relative_cm_subspace2,relative_cm_subspace1,
//            relative_space,relative_sectors,relative_matrices,J0
//          );
//        // std::cout << "As " << As << std::endl;
//
//        // construct M2
//        // PLACEHOLDER: Eigen::MatrixXd M2 = Eigen::MatrixXd::Zero(relative_cm_subspace2.size(),two_body_subspace2.size());
//        // TO UPDATE: Eigen::MatrixXd M2 = MoshinskyMatrix(relative_cm_subspace2,two_body_subspace2);
//        // std::cout << "M2 " << M2 << std::endl;
//
//        // construct M1
//        // PLACEHOLDER: Eigen::MatrixXd M1 = Eigen::MatrixXd::Zero(relative_cm_subspace1.size(),two_body_subspace1.size());
//        // TO UPDATE: Eigen::MatrixXd M1 = MoshinskyMatrix(relative_cm_subspace1,two_body_subspace1);
//        // std::cout << "M1 " << M1 << std::endl;
//      
//        // generate contribution to A
//        Eigen::MatrixXd A_contribution = 2 * M2.transpose() * As * M1;
//
//        A += A_contribution;
//      }


  return A;

}


  ////////////////////////////////////////////////////////////////
  ////////////////////////////////////////////////////////////////
} // namespace
