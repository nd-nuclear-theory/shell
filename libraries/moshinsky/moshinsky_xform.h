/****************************************************************
  moshinsky_xform.h

  Defines Moshinsky transformation of general relative operator in
  LSJT scheme.

  See notes "Moshinsky xform for operators" (2015) for derivation.

  Language: C++11
                                 
  Mark A. Caprio
  University of Notre Dame

  11/13/15 (mac): Created (as program moshinsky.cpp).
  11/21/15 (mac): Extract interaction I/O and manipulation functions
    to interaction_lsjt.  Allow input of generic relative interaction.
  11/26/15 (mac): Initial running version.
  1/6/16 (mac): Documentation updates.
  7/4/16 (mac): Overhaul:
    - Revise for restructured shell directory structure.
    - Revise to use block-N structure, with updated lsjt_scheme indexing.
    - Update to group theory Wigner-Eckart convention.
    - Extract transformation routines from program file into header.

****************************************************************/

#ifndef MOSHINSKY_XFORM_H_
#define MOSHINSKY_XFORM_H_

#include "am/wigner_gsl.h"
#include "basis/lsjt_scheme.h"
#include "basis/lsjt_operator.h"

namespace moshinsky {

////////////////////////////////////////////////////////////////
// Racah reduction factor
////////////////////////////////////////////////////////////////

double RacahReductionFactorFirstSystemGT(
    const HalfInt& J1p, const HalfInt& Jp, 
    const HalfInt& J1, const HalfInt& J, 
    const HalfInt& J2, const HalfInt& J0
  );
// Prefactor for first-system operator in Racah two-system 
// reduction formula.
//
// Follows group theoretic (Rose) conventions for RME normalization:
//
//   <J1',J2;J||A_1^J0||J1,J2;J>
//     =(-)^(J1'+J2+J+J0)*Hat(J1')*Hat(J)*{J1',J',J2;J,J1,J0} 
//
// Note: Under Edmonds convention, the Hat(J1') would be a Hat(J').
//
// Arguments (all HalfInt):
//   J1p, Jp: bra first-system and total a.m.
//   J1, J: ket first-system and total a.m.
//   J2: bra/ket second-system a.m.
//   J0: operator a.m.
//   

////////////////////////////////////////////////////////////////
// Moshinsky transformation
////////////////////////////////////////////////////////////////

Eigen::MatrixXd
MoshinskyMatrixNLSJT(
    const basis::RelativeCMSubspaceNLSJT& relative_cm_subspace,
    const basis::TwoBodySubspaceNLSJT& two_body_subspace
  );
// Generate Moshinsky transformation matrix in given NLSJT sector.
//
// The result consists of Moshinsky brackets.
//
// The spectator SJT labels for the subspaces are assumed to be identical.
//
// Arguments:
//   relative_cm_subspace (basis::RelativeCMSubspaceNLSJT) : the relative-cm subspace
//   two_body_subspace (basis::TwoBodySubspaceNLSJT) : the two-body subspace
//
// Returns:
//   (matrix) : the transformation brackets


Eigen::MatrixXd
RelativeCMMatrixNLSJT(
    const basis::RelativeSpaceLSJT& relative_space,
    const basis::RelativeSectorsLSJT& relative_sectors,
    const basis::MatrixVector& relative_matrices,
    const typename basis::RelativeCMSectorsNLSJT::SectorType& relative_cm_sector,
    int J0
  );
// Generate JT-reduced matrix elements of an operator A^(J0,T0,g0) on
// a single fixed-N relative-cm sector (NLSJT), from the relative
// JT-reduced matrix elements.
//
// Since we are dealing with JT-reduced matrix elements, we work with
// each individual isospin component (T0) of the operator separately.
//
// This function implements the relation
//
//   <Nr',lr',Nc',lc';L',S',J',T',g'||A||Nr,lr,Nc,lc;L,S,J,T,g>
//     = Z(lr',lc,S';L',Jr';J') * Z(lr,lc,S;L,J';J)
//       * R1(Jr',lc,J';Jr,lc,J;J0)
//       *  <Nr';lr',S',Jr',T',gr'||A_r||Nr;lr,S,Jr,T,gr>
//
//  with a delta on (Nc',lc')=(Nc,lc), where the Z factors are
//  (12)3-(13)2 unitary recoupling coefficients, and R1 is the Racach
//  reduction formula factor for a spherical tensor operator which
//  only acts on the "first system".
//
// Arguments:
//   relative_space (basis::RelativeSpaceLSJT) : relative space
//     on which relative operator is defined
//   relative_sectors (basis::RelativeSectorsLSJT) : source sectors 
//     on which relative operator is defined
//   relative_matrices (basis::MatrixVector) : source matrix elements 
//     defining relative operator
//   relative_cm_sector (basis::RelativeCMSectorsNLSJT::SectorType) : target sector
//     for relative-cm operator
//   J0 (int) : the angular momentum carried by the operator (needed
//     in recoupling coefficient)
//
// Returns:
//   (Eigen::MatrixXd) : the matrix representation of this sector

Eigen::MatrixXd 
TransformedSector(
    const basis::TwoBodySubspaceLSJT& two_body_subspace2,
    const basis::TwoBodySubspaceLSJT& two_body_subspace1,
    const basis::RelativeSpaceLSJT& relative_space,
    const basis::RelativeSectorsLSJT& relative_sectors,
    const basis::MatrixVector& relative_matrices,
    int J0
  );
// Carries out Moshinsky transform to generate the reduced matrix
// elements on a LSTJ-coupled two-body sector.  
// 
// Important: The input relative matrices must contain *reduced*
// matrix elements, not plain matrix elements.  The present code is
// oriented towards the tranformation of general spherical tensor
// relative operators, for which it is appropriate to work with
// reduced metrix elements.  In contrast, for simple scalar
// (interaction-type) relative operators, common convention is to
// specify the relative interaction in terms of its plain, unreduced
// matrix elements (which are M-independent) in the relative subspace.
//
// The output matrix element are antisymmetrized (AS) matrix elements,
// rather than normalized antisymmetrized (NAS) matrix elements (on
// the one hand) or distinguishable-particle matrix elements (on the
// other).
//
// Note inclusion of factor of 2 relative two "naive" application of
// Moshinsky transform on distinguishable-particle states, arising
// from the bookkeeping since our two-body states are antisymmetrized,
// rather than distinguishable-particle states.


  ////////////////////////////////////////////////////////////////
  ////////////////////////////////////////////////////////////////
} // namespace

#endif
