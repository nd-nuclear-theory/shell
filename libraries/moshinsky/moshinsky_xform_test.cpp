/****************************************************************
  moshinsky_xform_test.cpp

  Mark A. Caprio
  University of Notre Dame

****************************************************************/

#include <iomanip>

#include "moshinsky/moshinsky_xform.h"

void test_moshinsky_matrix()
{

  std::cout << "Inspecting a Moshinsky matrix" << std::endl;

  // set up sectors
  int L=2;
  int S=1;
  int J=2;
  int T=0;
  int g=0;
  int N=4;

  std::cout << "  relative-cm" << std::endl;
  basis::RelativeCMSubspaceNLSJT relative_cm_subspace(L,S,J,T,g,N);
  std::cout << relative_cm_subspace.DebugStr();

  std::cout << "  two-body" << std::endl;
  basis::TwoBodySubspaceNLSJT two_body_subspace (L,S,J,T,g,N);
  std::cout << two_body_subspace.DebugStr();

  std::cout << "  Moshinsky matrix" << std::endl;
  Eigen::MatrixXd matrix = moshinsky::MoshinskyMatrixNLSJT(relative_cm_subspace,two_body_subspace);
  std::cout << matrix << std::endl;

  std::cout << "  Orthogonality test" << std::endl;
  std::cout << matrix.transpose()*matrix << std::endl;

}

void test_transform_isoscalar()
{

  ////////////////////////////////////////////////////////////////
  // define relative identity operator
  ////////////////////////////////////////////////////////////////

  std::cout << "Setup" << std::endl;

  // set up space
  int Nmax = 2;
  int Jmax = Nmax+1;
  basis::RelativeSpaceLSJT space(Nmax,Jmax);

  // set up operator containers
  //
  // These are vectors to store information for T0=0/1/2 components.
  std::vector<basis::RelativeSectorsLSJT> component_sectors(3);
  std::vector<basis::MatrixVector> component_matrices(3);

  // populate operator containers
  int J0 = 0;
  int g0 = 0;
  for (int T0=0; T0<=2; ++T0)
    // for each isospin component
    {

      // enumerate sectors
      component_sectors[T0] = basis::RelativeSectorsLSJT(space,J0,T0,g0);
      std::cout << " T0 " << T0 << " size " << component_sectors[T0].size() << std::endl;
          
      // populate matrices
      if (T0==0)
        basis::SetOperatorToIdentity(component_sectors[T0],component_matrices[T0]);
      else
        basis::SetOperatorToZero(component_sectors[T0],component_matrices[T0]);
    }

  ////////////////////////////////////////////////////////////////
  // augment to relative-cm
  ////////////////////////////////////////////////////////////////

  // // configuration parameters
  // const int Nmax_relative = 2;
  // const int J0 = 0;
  // const int g0 = 0;
  // const int Nmax_two_body = 2;

  // // set up relative space
  // std::cout << std::endl;
  // std::cout << "Setting up relative space..." << std::endl;
  // std::cout << "Nmax " << Nmax_relative << std::endl;
  // RelativeSpaceLSJT relative_space(Nmax_relative);
  // // WriteRelativeSubspaces(std::cout,relative_space);

  // // set up relative operator
  // std::cout << std::endl;
  // std::cout << "Setting up relative sectors..." << std::endl;
  // std::cout << "J0 " << J0 << " g0 " << g0 << std::endl;
  // RelativeSectorsLSJT relative_sectors(relative_space,J0,g0);  // for scalar operator
  // SectorMatrices Ar_matrices;
  // std::cout << std::endl;

  // // input relative operator
  // std::cout << "Reading relative operator..." << std::endl;
  // // ReadRelativeOperator(std::cin,relative_space,relative_sectors,Ar_matrices);
  // SetOperatorToIdentityReduced(relative_space,relative_sectors,Ar_matrices);

  // // output relative operator
  // std::cout << std::endl;
  // std::cout << "Relative operator" << std::endl;
  // // caution: affects future output precision
  // WriteRelativeOperatorMatrices(std::cout,relative_space,relative_sectors,Ar_matrices,10,7);

  // // set up two-body space
  // std::cout << std::endl;
  // std::cout << "Setting up two-body space..." << std::endl;
  // TwoBodySpaceLSJT two_body_space(Nmax_two_body);
  // std::cout << "Nmax " << Nmax_two_body << std::endl;
  // // WriteRelativeSubspaces(std::cout,relative_space);

  // // construct two-body operator
  // std::cout << std::endl;
  // std::cout << "Transforming..." << std::endl;
  // TwoBodySectorsLSJT two_body_sectors(two_body_space,J0,g0);  // for scalar operator
  // SectorMatrices A_matrices;
  // for (int two_body_sector_index = 0; two_body_sector_index < two_body_sectors.size(); ++two_body_sector_index)
  //   {
  //     Sector two_body_sector = two_body_sectors.GetSector(two_body_sector_index);
  //     const TwoBodySubspaceLSJT& two_body_subspace2 = two_body_space.GetSubspace(two_body_sector.index2());
  //     const TwoBodySubspaceLSJT& two_body_subspace1 = two_body_space.GetSubspace(two_body_sector.index1());
  //     Eigen::MatrixXd two_body_sector_matrix = TransformedSector(
  //         two_body_subspace2,two_body_subspace1,
  //         relative_space,relative_sectors,Ar_matrices,
  //         J0 // for scalar operator
  //       );

  //     A_matrices.push_back(two_body_sector_matrix);
  //   }

  // // dump two_body operator
  // std::cout << std::endl;
  // std::cout << "Two-body operator" << std::endl;
  // WriteTwoBodyOperator(std::cout,two_body_space,two_body_sectors,A_matrices);
  // WriteTwoBodyOperatorMatrices(std::cout,two_body_space,two_body_sectors,A_matrices,10,7);

  
}

////////////////////////////////////////////////////////////////
// main
////////////////////////////////////////////////////////////////

int main(int argc, char **argv)
{

  test_moshinsky_matrix();
  // test_transform_isoscalar();

  // termination
  return 0;
}
