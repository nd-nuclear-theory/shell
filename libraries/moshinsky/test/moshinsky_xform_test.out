Inspecting a Moshinsky matrix
  relative-cm
 index   0 Nr lr Nc lc   0   0   4   2
 index   1 Nr lr Nc lc   2   0   2   2
 index   2 Nr lr Nc lc   2   2   2   0
 index   3 Nr lr Nc lc   2   2   2   2
 index   4 Nr lr Nc lc   4   2   0   0
  two-body
 index   0 N1 l1 N2 l2   0   0   4   2
 index   1 N1 l1 N2 l2   1   1   3   1
 index   2 N1 l1 N2 l2   1   1   3   3
 index   3 N1 l1 N2 l2   2   0   2   2
 index   4 N1 l1 N2 l2   2   2   2   2
  Moshinsky matrix
        0.25      0.41833     0.273861     0.381881     0.288675
    0.381881     0.273861     -0.41833   -0.0833333    -0.440959
    0.381881    -0.273861      0.41833   -0.0833333    -0.440959
    0.288675 -1.10636e-17  9.76057e-17    -0.440959     0.666667
        0.25     -0.41833    -0.273861     0.381881     0.288675
  Orthogonality test
         0.5  1.38778e-17  4.16334e-17  1.11022e-16            0
 1.38778e-17          0.5            0 -5.55112e-17 -1.38778e-17
 4.16334e-17            0          0.5 -6.93889e-17  6.93889e-17
 1.11022e-16 -5.55112e-17 -6.93889e-17          0.5  1.38778e-17
           0 -1.38778e-17  6.93889e-17  1.38778e-17            1
